### PR TITLE
New: `no-const-assign` rule (Fixes #2719)

### LINKED ISSUE #2719: Rule Proposal: no-const-assign
Variables that declared with `const` cannot re-assign.
Under strict mode, it raise a runtime error. Otherwise just ignored (on V8).

The following patterns are considered warnings:

```js
const a = 0;
a = 1;
```

I'm happy to work if it was accepted.

### PR DESCRIPTION
**Pull request by mysticatea**
For the destructuring assignments, to check from `AssignmentExpression` or `VariableDeclaration` was very hard. So I traversed scopes.

### CODE REVIEW

File: docs/rules/no-const-assign.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+```js
+for (const a of [1, 2, 3]) { // `a` is re-defined (not modified) on each loop step.
+    console.log(a);
+}
+```
+
+## When Not to Use It
+
+If you don't want to be notified about modifying variables that are declared using `const` keyword, you can safely disable this rule.
```
Comment (mysticatea) [inline docs/rules/no-const-assign.md]: safely?

File: tests/lib/rules/no-const-assign.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            ecmaFeatures: {blockBindings: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "const x = 0; ++x;",
+            ecmaFeatures: {blockBindings: true},
+            errors: [{message: "`x` is constant.", type: "Identifier"}]
+        },
+        {
+            code: "for (const i = 0; i < 10; ++i) { foo(i); }",
```
Comment (gyandeeps) [inline tests/lib/rules/no-const-assign.js]: why is this an issue?

File: lib/rules/no-const-assign.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        var identifier = reference.identifier;
+        var range = identifier.range;
+        var traceKey = range[0] + "-" + range[1];
+
+        if (identifier != null &&
+            reference.init !== true &&
+            reference.isWrite() &&
+            // Destructuring assignments have multiple default value,
+            // so possibly there are multiple writeable references for the same identifier :(
+            traceMap[traceKey] !== true
```
Comment (gyandeeps) [inline lib/rules/no-const-assign.js]: Can this be re-written like this
```js
if (!identifier &&
     !reference.init &&
     reference.isWrite() &&
     // Destructuring assignments have multiple default value,
     // so possibly there are multiple writeable references for the same identifier :(
     !traceMap[traceKey]
}
```

File: tests/lib/rules/no-const-assign.js (hunk 2 above)
Comment (michaelficarra) [inline tests/lib/rules/no-const-assign.js]: It's only an issue if `++i` gets evaluated. So it's *potentially* an issue, depending on the implementation of `foo`.

**Comment by gyandeeps** [comment]
I would add some more test like 

```js
const a = 1;
function foo(){
 a = a + 1;
}
```

```js
const a = 1;
function foo(y){
 a = y;
}
```

```js
const a = 1;
while (true){
 a = a + 1;
}
```

```js
const myObject = {"key": "value"};
myObject.key = "otherValue";
```

**Comment by michaelficarra** [comment]
Is there no simpler way to check that a variable was declared as non-writable? If this isn't built into escope, we should add it.

**Comment by gyandeeps** [comment]
@michaelficarra I was kinda of thinking along the same line.
:+1:

File: tests/lib/rules/no-const-assign.js (hunk 2 above)
Comment (mysticatea) [inline tests/lib/rules/no-const-assign.js]: The `ForStatement.update` is evaluated normally, so this `++i` raises a runtime error.

Comment (mysticatea) [inline tests/lib/rules/no-const-assign.js]: `for-in`/`for-of` treats its loop variable specially in background if the variable is declared `let`/`const`.
In that case, `for-in`/`for-of` redefine its loop variable on each loop step, not modified.

On V8, this code is valid under strict mode:

```js
for (const a in [1,2,3]) { console.log(a); }
for (const a of [1,2,3]) { console.log(a); }
```

But this code raises a runtime error.

```js
for (const i = 0; i < 3; ++i) { console.log(i); }
```

Reference: 

- [ForIn/OfBodyEvaluation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-forin-div-ofbodyevaluation-lhs-stmt-iterator-lhskind-labelset) - 5. b. iv.
- [BindingInstantiation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-bindinginstantiation)

**Comment by mysticatea** [comment]
@michaelficarra Probably yes.
References of destructuring assignments can be multiple, it's a trouble maker. But I think, to merge those without breaking changes is difficult...

I have not sent yet (spec may be changed... but because I wanted to review for direction of the code, I sent this PR), `no-class-assign`'s code is similar this.  Now, `no-func-assign` has a bug about destructing assignments, so probably it becomes similar to this.
If we add new feature into escope, we should do now...

Sorry, I will sleep (AM 0:30 JST), ... I will add tests tomorrow.

**Comment by michaelficarra** [comment]
@gyandeeps your last example is invalid. That's a mutation, not assignment. Constant variables can't be re-assigned, but they can be mutated.

**Comment by gyandeeps** [comment]
@michaelficarra Last example is to make sure this rule doesn't warn. it should pass.

**Comment by nzakas** [comment]
Agree with @michaelficarra let's see if we can get this info into escope

**Comment by mysticatea** [comment]
I added tests, and minor fix. The diff is here: https://github.com/mysticatea/eslint/commit/6b97b06f28e1725ebd148f91eb45d8974b6df776

Umm? The CI test failed... `sinnon.sandbox`?

**Comment by mysticatea** [comment]
I simplified the `checkReference` function, `traceMap` is needless now.

I thought I want the common variable traversal system in ESLint.
Because there are many rules that traversing variables at `"Program:exit"` node.

e.g.

```js
module.exports = function(context) {
  // ....

  return {
    "variable": function(variable) { ... }
  };
};

**Comment by mysticatea** [comment]
I guess [this issue](https://github.com/cjohansen/Sinon.JS/issues/761) is cause of failing CI test.

**Comment by mysticatea** [comment]
I thought, an idea.

Let AST nodes of statement have a `variables` property, the property provides an array of variables that the statement declares.

For example:

```js
let {a: [x, y, ...z], b: w} = foo;
//--------
return {
    "VariableDeclaration": function(node) {
        console.log(node.variables); // [x, y, z, w]
    }
};
```

```js
function foo({a: [x, y, ...z], b: w}, a) {
    let b = 0;
}
//--------
return {
    "FunctionDeclaration": function(node) {
        console.log(node.variables); // [foo, x, y, z, w, a]
    }
};
```

```js
try {
}
catch ({message}) {
}
//--------
return {
    "CatchClause": function(node) {
        console.log(node.variables); // [message]
    }
};
```

Let `VariableDeclaration`, `FunctionDeclaration`, `FunctionExpression`, `CatchClause`, `ArrowFunctionExpression`, `ClassDeclaration`, `ClassExpression`, and `ImportDeclaration` have the `variable` property. Then we don't need to traverse all variables.

But now, `escope` does not modify given AST nodes. I think it's good design. So I don't know this idea is good or not.

Thought?

**Comment by nzakas** [comment]
We generally don't want to modify the AST nodes. You can do the same by using getScope(), I believe.

**Comment by mysticatea** [comment]
OK, I also prefer it.

```js
let {a: [x, y, ...z], b: w} = foo;
//--------
return {
    "VariableDeclaration": function(node) {
        var variables = context.getDeclaredVariables(node); // [x, y, z, w]
    }
};
```

This method returns variables that are declared by the node.
"are declared by the node" means the node is same as `Variable.defs[].node` or `Variable.defs[].parent`.
This method is `O(1)`.

I will try to add such a method to `escope::ScopeManager` and `eslint::RuleContext`.

**Comment by nzakas** [comment]
@mysticatea can you open an issue so we can discuss the use case?

**Comment by nzakas** [comment]
@mysticatea are you done with this PR?

**Comment by mysticatea** [comment]
I'm sorry, I opened new issue.

On escope, I'm implementing now.
If this spec is, probably I will complete in few hours.
After discuss, I will fix with the result, then send a PR to escope...

**Comment by mysticatea** [comment]
To think more simple way was separated to other issue, so I'm done with this.

**Comment by nzakas** [comment]
Awesome, thank you

**Comment by nzakas** [comment]
Just want to double-check on this PR, we are still waiting for escope?

**Comment by ilyavolodin** [comment]
@mysticatea are you still waiting for escope changes for this or can this be merged in (after rebase)?

**Comment by mysticatea** [comment]
Yeah, I'm waiting :(
I reminded Suzuki-san with Twitter 5-days ago.

**Comment by mysticatea** [comment]
I'm sorry to have kept you waiting.

I made the `no-const-assign` rule using `context.getDeclaredVariables(node)`.

**Comment by ilyavolodin** [comment]
@mysticatea No problem. Can you rebase your PR please?

**Comment by mysticatea** [comment]
I rebased.

**Pull request merged**